opal_client/tests/data_updater_test.py

```diff
@@ -100,14 +127,47 @@ async def test_data_updater(server):
     # config to use mock OPA
     policy_store = PolicyStoreClientFactory.create(store_type=PolicyStoreTypes.MOCK)
     updater = DataUpdater(pubsub_url=UPDATES_URL, policy_store=policy_store,
-                          fetch_on_connect=False, data_topics=DATA_TOPICS)
+                          fetch_on_connect=False, data_topics=DATA_TOPICS, should_send_reports=False)
+    # start the updater (terminate on exit)
+    await updater.start()
+    try:
+        proc = multiprocessing.Process(target=trigger_update, daemon=True)
+        proc.start()
+        # wait until new data arrives into the strore via the updater
+        await asyncio.wait_for(policy_store.wait_for_data(), 15)
+    # cleanup
+    finally:
+        await updater.stop()
+        proc.terminate()
+
+@pytest.mark.asyncio
+async def test_data_updater_with_report_callback(server):
+    """
+    Disable auto-update on connect (fetch_on_connect=False)
+    Connect to OPAL-server trigger a Data-update and check our policy store gets the update
+    """
+    # Wait for the server to start
+    server.wait(5)
+    # config to use mock OPA
+    policy_store = PolicyStoreClientFactory.create(store_type=PolicyStoreTypes.MOCK)
+    updater = DataUpdater(pubsub_url=UPDATES_URL, policy_store=policy_store,
+                          fetch_on_connect=False, data_topics=DATA_TOPICS, should_send_reports=True)
     # start the updater (terminate on exit)
     await updater.start()
     try:
         proc = multiprocessing.Process(target=trigger_update, daemon=True)
         proc.start()
         # wait until new data arrives into the strore via the updater
         await asyncio.wait_for(policy_store.wait_for_data(), 15)
+        # give the callback a chance to arrive
+        await asyncio.sleep(1)
+
+        async with ClientSession() as session:
+            res = await session.get(CHECK_DATA_UPDATE_CALLBACK_URL)
+            count = await res.json()
+            # we got one callback
+            assert count == 1
+
```

Comment:
i suspect this is flaky because in the CI environment scheduling of the deferred async tasks (Created by asyncio.create_task) takes a longer time.

Maybe there is a way to "force-await" in the test so that the task is actually called and scheduled?

a way to verify my assumption would be to log something in the task, if you see it in the log when the flaky test fails, then i am wrong :)

Reply:
There's no indication this is flakey